File: Solutions/332. Reconstruct Itinerary.cpp
#include <iostream>
#include <vector>
#include <set>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <multiset>

using namespace std;

class Solution {
private:
    void DFS(unordered_map<string, multiset<string>>& AdjList,
             string current,
             vector<string>& itinerary) {

        while (!AdjList[current].empty()) {
            auto it = AdjList[current].begin();
            string next = *it;

            AdjList[current].erase(it);

            DFS(AdjList, next, itinerary);
        }

        itinerary.push_back(current);
    }

public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {

        unordered_map<string, multiset<string>> AdjList;

        for (const auto& ticket : tickets) {
            AdjList[ticket[0]].insert(ticket[1]);
        }

        vector<string> itinerary;

        DFS(AdjList, "JFK", itinerary);

        reverse(itinerary.begin(), itinerary.end());

        return itinerary;
    }
};

int main() {

    Solution sol;

    vector<vector<string>> tickets = {
        {"JFK","SFO"},
        {"JFK","ATL"},
        {"SFO","ATL"},
        {"ATL","JFK"},
        {"ATL","SFO"}
    };

    vector<string> itinerary = sol.findItinerary(tickets);

    for (const auto& city : itinerary) {
        cout << city << " ";
    }

    cout << endl;

    return 0;
}

File: Solutions/PY/332. Reconstruct Itinerary.py
from collections import defaultdict


class Solution:
    def DFS(self, adj_list, current, itinerary):

        while adj_list[current]:
            # smallest lexical destination
            next_city = adj_list[current][0]

            # remove only one occurrence
            adj_list[current].pop(0)

            self.DFS(adj_list, next_city, itinerary)

        itinerary.append(current)

    def findItinerary(self, tickets):

        adj_list = defaultdict(list)

        # build graph
        for src, dst in tickets:
            adj_list[src].append(dst)

        # keep lexical order like multiset
        for city in adj_list:
            adj_list[city].sort()

        itinerary = []

        self.DFS(adj_list, "JFK", itinerary)

        itinerary.reverse()

        return itinerary


if __name__ == "__main__":
    sol = Solution()

    tickets = [
        ["JFK", "SFO"],
        ["JFK", "ATL"],
        ["SFO", "ATL"],
        ["ATL", "JFK"],
        ["ATL", "SFO"],
    ]

    itinerary = sol.findItinerary(tickets)

    print(itinerary)
